src/main/scala/example/Service.scala

```diff
@@ -0,0 +1,131 @@
+package example
+
+import cats.{Id, Monad}
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+/**
+  * Repository and Service implementation using tagless final pattern.
+  * The idea is to make it easier to test our database layer, using Scala’s higher kinded types to abstract
+  * the Future type constructor away from our traits under test.
+  * Intro to tagless final: https://www.basementcrowd.com/2019/01/17/an-introduction-to-tagless-final-in-scala/.
+  * The similar task example https://github.com/LvivScalaClub/cats-playground/blob/master/src/main/scala/BookRepository.scala
+  */
+
+case class User(id: Long, username: String)
+case class IotDevice(id: Long, userId: Long, sn: String)
+
+// NOTE: This import bring into the scope implicits that allow you to call .map and .flatMap on the type F[_]
+// and also bring you typeclasses that know how to flatmap (Monad) and map (Functor) over your higher-kinded type.
+import cats.implicits._
+
+trait UserRepository[F[_]] {
+  def registerUser(username: String): F[User]
+  def getById(id: Long): F[Option[User]]
+  def getByUsername(username: String): F[Option[User]]
+}
+
+class UserRepositoryId extends UserRepository[Id] {
+  private var storage: Map[Long, User] = Map()
+
+  override def registerUser(username: String): Id[User] = {
+    val existing = getByUsername(username)
+    existing.getOrElse {
+      val id = if(storage.nonEmpty) {storage.keys.last + 1} else 1
+      val user = User(id, username)
+      storage = storage + (id -> user)
+      user
+    }
+  }
+  override def getById(id: Long): Id[Option[User]] = storage.get(id)
+  override def getByUsername(username: String): Id[Option[User]] = storage.values.find(user => user.username == username)
+}
+
+class UserRepositoryFuture extends UserRepository[Future] {
```

Comment:
it is considered a good practice to have execution context as implicit param to that class so you can control later which EC executes your futures. Global EC should be used very carefully.